Update custom state injection and add `ImmutableState` class (#845)
starlite-api/starlite: docs/usage/0-the-starlite-app/2-using-application-state.md

```py title="Using Application State"
--8<-- "examples/using_application_state.py"
```

## Injecting Application State into Route Handlers and Dependencies

As seen in the above example, Starlite offers an easy way to inject state into route handlers and dependencies - simply
by specifying `state` as a kwarg to the handler function. I.e., you can simply do this in handler function or dependency
to access the application state:

```python
from starlite import get, State

@get("/")
def handler(state: State) -> None:
    ...
```

When using this pattern you can specify the class to use for the state object. This type is not merely for type
checkers, rather Starlite will instantiate a new state instance based on the type you set there. This allows users to
use custom classes for State, e.g.:

```python
from starlite import get, State

class MyState(State):
    count: int = 0

    def increment(self) -> None:
        self.count *= 1

@get("/")
def handler(state: MyState) -> dict:
    state.increment()
    return state.dict()
```

While this is very powerful, it might encourage users to follow anti-patterns: its important to emphasize that using

> state can lead to hard to reason about code as well difficult to understand bugs due to changes in different ASGI

Reviewer: ```suggestion
state can lead to code that's hard to reason about and bugs that are difficult to understand, due to changes in different ASGI
```
